TelemetryDesignDoc.md

```diff
@@ -0,0 +1,327 @@
+        }
+      ]
+    }
+  ]
+}
+```
+
+A simple builder API, similar to the current Shuffleboard API, can allow dashboard configuration from robot code:
+
+```java
+var exampleTab = Telemetry.getDashboardTab("Drive");
+exampleTab.addWidget(new Widgets.Gyro("Heading", "drive/heading"))
+  // Graph widget ctor can be variadic to handle multiple data sources
+  .addWidget(new Widgets.Graph("Wheel Speeds", "drive/leftVelocity", "drive/rightVelocity")
+    .withXLimits(0, 10)
+    .withYLimits(-50, 50));
+```
+
+The above differs from the Shuffleboard API in that it has direct programmatic representation for widgets in the form of the widget classes (e.g. `Widget.Gyro`).  Shuffleboard only exposes a generic string-valued `withProperties` decorator, which does not provide type safety for user-specified widget properties.
+
+### Deprecate all existing SmartDashboard and Shuffleboard methods, replace with implementation-neutral equivalents
+
+With an implementation-neutral metadata spec available, telemetry publishing is no longer tightly coupled to the user's choice of dashboard.  Accordingly, there is no need to maintain separate `SmartDashboard` and `Shuffleboard` APIs.  In fact, it is not necessary for the API shape itself to contain any reference to the dashboard implementation or spec - this will be done only where it is convenient and reduces boilerplate in user code.
+
+The unified replacement API will look something like:
+
+```java
+// Simple imperative logging directly delegates to NT4
+Telemetry.publishDouble("key", value);
+// Can support dimensions through the API by delegating to metadata (C++ could use templates and be smarter)
+Telemetry.publishQuantity("key", value, "meters");
+
+// Declarative binding with suppliers/consumers
+Telemetry.publishDouble("key", () -> value);
+Telemetry.subscribeDouble("key", (double value) -> doThing(value));
+
+// Declarative binding of mutable classes similar to `putData(Sendable)` in existing impl
+// Optional final metadata string for aggregation
+var telemetryBinding = Telemetry.bind("key", telemetryNode);
+// Unbind later if needed (for cleanup etc)
+telemetryBinding.unbind();
+```
+
+Metadata for a given topic will be deep-merged into the existing metadata as they are introduced.  There does not seem to be any need for a system to remove metadata - such an interaction would be a code smell.
+
+### Replace Sendable with new, cleaner API
+
+`Sendable` should be replaced by a new interface, `TelemetryNode`.  The API shape for this will be fundamentally similar to `Sendable`:
+
+```java
+public interface TelemetryNode {
+  void bind(TelemetryBuilder builder);
+}
+```
+
+The "bind" method fulfills the role of `initSendable`, with a few key differences:
+
+Unlike `Sendable`, both `TelemetryNode` will support hierarchical nesting of implementing objects - that is, in addition to exposing binding handles for fields of primitive data type, it will allow *entire* `TelemetryNode`s to be added as children:
```

Comment:
How is lifetime of the children managed in C++?  A core function of SendableRegistry operation and the reason SendableHelper exists in C++ is to handle object moves.